File: browser-extension/src/shared/greenhouseDetector.ts
/**
 * Read-only Greenhouse field detector (M1 TypeScript port of M0 logic).
 * Never reads input values, never mutates DOM, never submits.
 */

import {
  DETECTOR_VERSION,
  EXTENSION_VERSION,
  type DetectedField,
  type FieldCategory,
  type FormAnalysisResult,
} from "../types/messages";
import { extractBoardToken, isSupportedGreenhouseUrl } from "../utils/url";

const SUPPORTED_NORMALIZED = new Set([
  "first_name", "last_name", "full_name", "email", "phone",
  "city", "state", "country", "postal_code",
  "linkedin_url", "portfolio_url", "github_url",
  "current_company", "current_title",
  "resume_upload", "cover_letter_upload",
  "work_authorization", "sponsorship_required",
]);

const SENSITIVE_PATTERNS = [
  /gender/i, /race/i, /ethnic/i, /hispanic/i, /disability/i, /disabled/i,
  /veteran/i, /military/i, /criminal/i, /conviction/i, /arrest/i,
  /medical/i, /\bhealth\b/i, /lgbt/i, /sexual orientation/i, /religion/i,
  /self[- ]identif/i, /eeo/i, /equal employment/i, /voluntary/i, /demographic/i,
];

const LEGAL_PATTERNS = [
  /certify/i, /attest/i, /acknowledge/i, /consent/i, /gdpr/i,
  /terms of/i, /privacy policy/i, /truthful/i, /accurate/i,
  /background check/i, /authorize .* check/i,
];

const WORK_AUTH_PATTERNS = [
  /work authorization/i, /authorized to work/i, /legally authorized/i,
  /eligible to work/i, /right to work/i,
];

const SPONSOR_PATTERNS = [/sponsor/i, /visa/i, /h-?1b/i, /require sponsorship/i];

const LABEL_MAP: [RegExp, string][] = [
  [/^first\s*name$/i, "first_name"],
  [/^last\s*name$/i, "last_name"],
  [/^full\s*name$|^name$/i, "full_name"],
  [/^e-?mail/i, "email"],
  [/^phone|^mobile|^cell/i, "phone"],
  [/^city$/i, "city"],
  [/^state|^province/i, "state"],
  [/^country/i, "country"],
  [/^zip|^postal/i, "postal_code"],
  [/linkedin/i, "linkedin_url"],
  [/portfolio|personal\s*website|website/i, "portfolio_url"],
  [/github/i, "github_url"],
  [/current\s*company|company\s*name/i, "current_company"],
  [/current\s*(job\s*)?title|job\s*title/i, "current_title"],
  [/^resume|^cv\b|curriculum/i, "resume_upload"],
  [/cover\s*letter/i, "cover_letter_upload"],
];

export function normalizeFieldLabel(
  label: string,
  fieldType = "text",
): { normalized: string | null; classification: FieldCategory; reason: string; confidence: number } {
  const raw = (label || "").trim();
  const low = raw.toLowerCase().replace(/[*:]+$/g, "").trim();

  if (SENSITIVE_PATTERNS.some((p) => p.test(low))) {
    return { normalized: null, classification: "sensitive_question", reason: "sensitive_pattern", confidence: 0.9 };
  }
  if (LEGAL_PATTERNS.some((p) => p.test(low))) {
    return { normalized: null, classification: "legal_attestation", reason: "legal_pattern", confidence: 0.85 };
  }
  if (SPONSOR_PATTERNS.some((p) => p.test(low))) {
    return { normalized: "sponsorship_required", classification: "supported", reason: "sponsor_pattern", confidence: 0.88 };
  }
  if (WORK_AUTH_PATTERNS.some((p) => p.test(low))) {
    return { normalized: "work_authorization", classification: "supported", reason: "work_auth_pattern", confidence: 0.88 };
  }

  if (fieldType === "file" || /resume/i.test(low) || low === "cv") {
    if (/cover/i.test(low)) {
      return { normalized: "cover_letter_upload", classification: "supported", reason: "file_cover", confidence: 0.95 };
    }
    if (/resume/i.test(low) || low === "cv" || /curriculum/i.test(low)) {
      return { normalized: "resume_upload", classification: "supported", reason: "file_resume", confidence: 0.95 };
    }
  }

  for (const [pattern, name] of LABEL_MAP) {
    if (pattern.test(low) && SUPPORTED_NORMALIZED.has(name)) {
      return { normalized: name, classification: "supported", reason: `label_map:${name}`, confidence: 0.92 };
    }
  }

  if (/(salary|compensation|clearance|security clearance)/i.test(low)) {
    return { normalized: null, classification: "unsupported", reason: "unsupported_bucket", confidence: 0.8 };
  }
  if (raw) {
    return { normalized: null, classification: "custom_question", reason: "unmapped_label", confidence: 0.7 };
  }
  return { normalized: null, classification: "unknown", reason: "empty_label", confidence: 0.4 };
}

function classifyDetected(
  label: string,
  fieldType: string,
  name: string,
  opts?: { verbose?: boolean; selector?: string },
): DetectedField {
  const { normalized, classification, reason, confidence } = normalizeFieldLabel(label || name, fieldType);
  const isUpload =
    fieldType === "file" ||
    normalized === "resume_upload" ||
    normalized === "cover_letter_upload";
  const field: DetectedField = {
    external_field_key: name || label || "unnamed",
    field_label: label || name || "",
    field_type: fieldType,
    is_required: false,
    is_upload: isUpload,
    normalized_field_name: normalized,
    classification,
    options: [],
    confidence,
  };
  if (opts?.verbose) {
    field.detection_reason = reason;
    if (opts.selector) field.matched_selector = opts.selector;
  }
  return field;
}

function finalize(
  partial: Omit<FormAnalysisResult, "supported_fields" | "sensitive_fields" | "custom_fields" | "legal_fields" | "upload_controls" | "required_fields" | "unsupported_fields" | "filled_any_fields" | "submitted" | "detector_version" | "extension_version" | "analyzed_at" | "page_mutated" | "warnings"> & {
    warnings?: string[];
  },
): FormAnalysisResult {
  const result: FormAnalysisResult = {
    ...partial,
    warnings: partial.warnings ?? [],
    supported_fields: [],
    sensitive_fields: [],
    custom_fields: [],
    legal_fields: [],
    upload_controls: [],
    required_fields: [],
    unsupported_fields: [],
    filled_any_fields: false,
    submitted: false,
    page_mutated: false,
    detector_version: DETECTOR_VERSION,
    extension_version: EXTENSION_VERSION,
    analyzed_at: new Date().toISOString(),
  };

  for (const f of result.fields) {
    if (f.classification === "supported" && f.normalized_field_name) {
      if (!result.supported_fields.includes(f.normalized_field_name)) {
        result.supported_fields.push(f.normalized_field_name);
      }
    } else if (f.classification === "sensitive_question") {
      result.sensitive_fields.push(f.field_label || f.external_field_key);
    } else if (f.classification === "custom_question") {
      result.custom_fields.push(f.field_label || f.external_field_key);
    } else if (f.classification === "legal_attestation") {
      result.legal_fields.push(f.field_label || f.external_field_key);
    } else if (f.classification === "unsupported" || f.classification === "unknown") {
      result.unsupported_fields.push(f.field_label || f.external_field_key);
    }
    if (f.is_upload) result.upload_controls.push(f.field_label || f.external_field_key);
    if (f.is_required) result.required_fields.push(f.field_label || f.external_field_key);
  }
  if (!result.is_greenhouse) {
    result.warnings.push("Page/URL does not look like Greenhouse");
  }
  return result;
}

export function detectGreenhouseUrl(url: string | null | undefined): boolean {
  return isSupportedGreenhouseUrl(url);
}

/** Parse HTML string (fixtures / offline). Does not execute scripts. */
export function detectFromHtml(
  html: string,
  opts?: { applicationUrl?: string | null; verbose?: boolean },
): FormAnalysisResult {
  const doc = new DOMParser().parseFromString(html || "", "text/html");
  return detectFromDocument(doc, {
    applicationUrl: opts?.applicationUrl ?? null,
    verbose: opts?.verbose,
  });
}

export function detectFromDocument(
  doc: Document,
  opts?: { applicationUrl?: string | null; verbose?: boolean },
): FormAnalysisResult {
  const applicationUrl = opts?.applicationUrl ?? null;
  const urlIsGh = detectGreenhouseUrl(applicationUrl);
  const htmlText = doc.documentElement?.outerHTML?.toLowerCase() ?? "";
  const hasGrnhse =
    !!doc.getElementById("grnhse_app") ||
    !!doc.querySelector('iframe[src*="greenhouse"]') ||
    htmlText.includes("greenhouse") ||
    htmlText.includes("grnhse");
  const isGh = urlIsGh || hasGrnhse;

  let jobTitle =
    doc.querySelector("h1")?.textContent?.trim() ||
    doc.querySelector('meta[property="og:title"]')?.getAttribute("content")?.trim() ||
    doc.title?.trim() ||
    null;
  let employer: string | null = null;
  if (jobTitle && / at /i.test(jobTitle)) {
    const parts = jobTitle.split(/\s+at\s+/i);
    if (parts.length >= 2) {
      jobTitle = parts[0].trim();
      employer = parts.slice(1).join(" at ").trim();
    }
  }
  const board = extractBoardToken(applicationUrl);
  if (!employer && board && board !== "embed") employer = board;

  const labelFor = new Map<string, string>();
  doc.querySelectorAll("label[for]").forEach((lab) => {
    const f = lab.getAttribute("for");
    if (f) labelFor.set(f, (lab.textContent || "").trim());
  });

  const fields: DetectedField[] = [];
  const controls = doc.querySelectorAll("input, textarea, select");
  controls.forEach((el) => {
    const tag = el.tagName.toLowerCase();
    const input = el as HTMLInputElement;
    const type = (input.type || tag).toLowerCase();
    if (type === "hidden" || type === "submit" || type === "button") return;

    // CRITICAL: never read .value / checked / selected — labels & structure only
    const name = input.name || input.id || "";
    const id = input.id || "";
    let label = (id && labelFor.get(id)) || "";
    if (!label) {
      const wrapping = el.closest("label");
      if (wrapping) label = (wrapping.textContent || "").trim();
    }
    const legend = el.closest("fieldset")?.querySelector("legend")?.textContent?.trim() || "";
    if (legend && (!label || ["yes", "no", name.toLowerCase()].includes(label.toLowerCase()))) {
      label = legend;
    }
    if (!label && name) label = name.replace(/_/g, " ");

    const required =
      input.required ||
      input.getAttribute("aria-required") === "true" ||
      !!el.closest("[aria-required='true']");

    const selector = opts?.verbose
      ? `${tag}${id ? `#${CSS.escape(id)}` : name ? `[name="${name}"]` : ""}`
      : undefined;

    const df = classifyDetected(label, type, name, { verbose: opts?.verbose, selector });
    df.is_required = required;

    if (tag === "select") {
      const optsList: string[] = [];
      (el as HTMLSelectElement).querySelectorAll("option").forEach((o) => {
        const t = (o.textContent || "").trim();
        // option text is structural (choices), not user-entered answers
        if (t && optsList.length < 50) optsList.push(t);
      });
      df.options = optsList;
    }

    fields.push(df);
  });

  let confidence = 0.5;
  if (urlIsGh && hasGrnhse) confidence = 0.98;
  else if (urlIsGh) confidence = 0.9;
  else if (hasGrnhse && fields.length > 0) confidence = 0.85;
  else if (hasGrnhse) confidence = 0.7;

  return finalize({
    is_greenhouse: isGh,
    platform: isGh ? "greenhouse" : "unknown",
    confidence,
    detection_mode: "dom",
    employer,
    job_title: jobTitle,
    application_url: applicationUrl,
    board_token: board,
    fields,
  });
}

/**
 * Snapshot structural fingerprints before/after analysis to prove no mutation.
 * Does not include field values.
 */
export function structuralFingerprint(doc: Document): string {
  const parts: string[] = [];
  doc.querySelectorAll("input, textarea, select").forEach((el) => {
    const input = el as HTMLInputElement;
    parts.push(
      [
        el.tagName,
        input.name || "",
        input.id || "",
        input.type || "",
        input.required ? "1" : "0",
        // Explicitly omit value/checked/selectedIndex/files
      ].join(":"),
    );
  });
  return parts.join("|");
}

File: backend/services/greenhouse_detector.py
"""Read-only Greenhouse application form detector (Phase 5 M0).

May inspect HTML / Job Board question JSON and return diagnostics.
Must never fill fields, upload files, click buttons, or submit.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass, asdict, field
from html.parser import HTMLParser
from pathlib import Path
from typing import Any, Optional

from services.application_url import (
    PLATFORM_GREENHOUSE,
    normalize_application_url,
)

# M0 normalized profile field keys we may map labels to.
SUPPORTED_NORMALIZED = frozenset({
    "first_name", "last_name", "full_name", "email", "phone",
    "city", "state", "country", "postal_code",
    "linkedin_url", "portfolio_url", "github_url",
    "current_company", "current_title",
    "resume_upload", "cover_letter_upload",
    "work_authorization", "sponsorship_required",
})

SENSITIVE_PATTERNS = (
    r"gender", r"race", r"ethnic", r"hispanic", r"disability", r"disabled",
    r"veteran", r"military", r"criminal", r"conviction", r"arrest",
    r"medical", r"health\b", r"lgbt", r"sexual orientation", r"religion",
    r"self[- ]identif", r"eeo", r"equal employment", r"voluntary",
    r"demographic",
)

LEGAL_PATTERNS = (
    r"certify", r"attest", r"acknowledge", r"consent", r"gdpr",
    r"terms of", r"privacy policy", r"truthful", r"accurate",
    r"background check", r"authorize .* check",
)

WORK_AUTH_PATTERNS = (
    r"work authorization", r"authorized to work", r"legally authorized",
    r"eligible to work", r"right to work",
)

SPONSOR_PATTERNS = (
    r"sponsor", r"visa", r"h-?1b", r"require sponsorship",
)

LABEL_MAP = [
    (r"^first\s*name$", "first_name"),
    (r"^last\s*name$", "last_name"),
    (r"^full\s*name$|^name$", "full_name"),
    (r"^e-?mail", "email"),
    (r"^phone|^mobile|^cell", "phone"),
    (r"^city$", "city"),
    (r"^state|^province", "state"),
    (r"^country", "country"),
    (r"^zip|^postal", "postal_code"),
    (r"linkedin", "linkedin_url"),
    (r"portfolio|personal\s*website|website", "portfolio_url"),
    (r"github", "github_url"),
    (r"current\s*company|company\s*name", "current_company"),
    (r"current\s*(job\s*)?title|job\s*title", "current_title"),
    (r"^resume|^cv\b|curriculum", "resume_upload"),
    (r"cover\s*letter", "cover_letter_upload"),
]


@dataclass
class DetectedField:
    external_field_key: str
    field_label: str
    field_type: str
    is_required: bool = False
    is_upload: bool = False
    normalized_field_name: Optional[str] = None
    classification: str = "unknown"  # supported | custom_question | sensitive_question | legal_attestation | unsupported | unknown
    options: list[str] = field(default_factory=list)


@dataclass
class GreenhouseDetectionResult:
    is_greenhouse: bool
    platform: str = PLATFORM_GREENHOUSE
    detection_mode: str = "none"  # html | greenhouse_api_json | url
    employer: Optional[str] = None
    job_title: Optional[str] = None
    application_url: Optional[str] = None
    board_token: Optional[str] = None
    fields: list[DetectedField] = field(default_factory=list)
    supported_fields: list[str] = field(default_factory=list)
    sensitive_fields: list[str] = field(default_factory=list)
    custom_fields: list[str] = field(default_factory=list)
    legal_fields: list[str] = field(default_factory=list)
    upload_controls: list[str] = field(default_factory=list)
    required_fields: list[str] = field(default_factory=list)
    unsupported_fields: list[str] = field(default_factory=list)
    warnings: list[str] = field(default_factory=list)
    filled_any_fields: bool = False  # always False in M0
    submitted: bool = False  # always False in M0

    def to_dict(self) -> dict:
        d = asdict(self)
        return d


class _FormHTMLParser(HTMLParser):
    def __init__(self):
        super().__init__()
        self.fields: list[dict] = []
        self._label_for: dict[str, str] = {}
        self._current_label_for: Optional[str] = None
        self._current_label_text: list[str] = []
        self._in_label = False
        self.meta: dict[str, str] = {}
        self.has_grnhse = False
        self.title_text: list[str] = []
        self._in_title = False
        self._in_h1 = False
        self.h1_text: list[str] = []
        self._fieldset_legend_stack: list[str] = []
        self._in_legend = False
        self._legend_text: list[str] = []

    def handle_starttag(self, tag, attrs):
        ad = dict(attrs)
        if tag == "title":
            self._in_title = True
        if tag == "h1":
            self._in_h1 = True
        if tag == "legend":
            self._in_legend = True
            self._legend_text = []
        if tag == "fieldset":
            self._fieldset_legend_stack.append("")
        if tag == "div" and ad.get("id") == "grnhse_app":
            self.has_grnhse = True
        if tag == "iframe" and "greenhouse" in (ad.get("src") or "").lower():
            self.has_grnhse = True
        if tag == "meta" and ad.get("property") == "og:title":
            self.meta["og:title"] = ad.get("content") or ""
        if tag == "label":
            self._in_label = True
            self._current_label_for = ad.get("for")
            self._current_label_text = []
        if tag in ("input", "textarea", "select"):
            name = ad.get("name") or ad.get("id") or ""
            ftype = ad.get("type") or tag
            if ftype == "hidden" or ftype == "submit" or ftype == "button":
                return
            required = "required" in ad or ad.get("aria-required") == "true"
            label = self._label_for.get(ad.get("id") or "", "") or name
            legend = self._fieldset_legend_stack[-1] if self._fieldset_legend_stack else ""
            if legend and (not label or label.lower() in ("yes", "no", name.lower())):
                label = legend
            self.fields.append({
                "name": name,
                "type": ftype,
                "required": required,
                "label": label,
                "accept": ad.get("accept"),
            })

    def handle_endtag(self, tag):
        if tag == "title":
            self._in_title = False
        if tag == "h1":
            self._in_h1 = False
        if tag == "legend" and self._in_legend:
            text = " ".join(self._legend_text).strip()
            if self._fieldset_legend_stack:
                self._fieldset_legend_stack[-1] = text
            self._in_legend = False
        if tag == "fieldset" and self._fieldset_legend_stack:
            self._fieldset_legend_stack.pop()
        if tag == "label" and self._in_label:
            text = " ".join(self._current_label_text).strip()
            if self._current_label_for:
                self._label_for[self._current_label_for] = text
            if text and self.fields and not self.fields[-1].get("label"):
                self.fields[-1]["label"] = text
            self._in_label = False

    def handle_data(self, data):
        if self._in_title:
            self.title_text.append(data)
        if self._in_h1:
            self.h1_text.append(data)
        if self._in_label:
            self._current_label_text.append(data)
        if self._in_legend:
            self._legend_text.append(data)


def _match_any(text: str, patterns: tuple[str, ...]) -> bool:
    t = text.lower()
    return any(re.search(p, t) for p in patterns)


def normalize_field_label(label: str, field_type: str = "text") -> tuple[Optional[str], str]:
    """Return (normalized_name|None, classification)."""
    raw = (label or "").strip()
    low = raw.lower().rstrip("*: ")

    if _match_any(low, SENSITIVE_PATTERNS):
        return None, "sensitive_question"
    if _match_any(low, LEGAL_PATTERNS):
        return None, "legal_attestation"
    if _match_any(low, SPONSOR_PATTERNS):
        return "sponsorship_required", "supported"
    if _match_any(low, WORK_AUTH_PATTERNS):
        return "work_authorization", "supported"

    if field_type in ("file", "input_file") or "resume" in low or low == "cv":
        if "cover" in low:
            return "cover_letter_upload", "supported"
        if "resume" in low or low == "cv" or "curriculum" in low:
            return "resume_upload", "supported"

    for pattern, name in LABEL_MAP:
        if re.search(pattern, low):
            if name in SUPPORTED_NORMALIZED:
                return name, "supported"

    # Known unsupported free-form buckets
    if any(x in low for x in ("salary", "compensation", "clearance", "security clearance")):
        return None, "unsupported"

    if raw:
        return None, "custom_question"
    return None, "unknown"


def _classify_detected(label: str, field_type: str, name: str) -> DetectedField:
    norm, classification = normalize_field_label(label or name, field_type)
    is_upload = field_type in ("file", "input_file") or norm in ("resume_upload", "cover_letter_upload")
    return DetectedField(
        external_field_key=name or label or "unnamed",
        field_label=label or name or "",
        field_type=field_type,
        is_upload=is_upload,
        normalized_field_name=norm,
        classification=classification,
    )


def detect_greenhouse_url(url: Optional[str]) -> bool:
    c = normalize_application_url(url)
    return c.is_valid and c.platform == PLATFORM_GREENHOUSE


def detect_from_html(html: str, *, application_url: Optional[str] = None) -> GreenhouseDetectionResult:
    parser = _FormHTMLParser()
    parser.feed(html or "")
    url_is_gh = detect_greenhouse_url(application_url) if application_url else False
    html_is_gh = (
        parser.has_grnhse
        or "greenhouse" in (html or "").lower()
        or "grnhse" in (html or "").lower()
    )
    is_gh = url_is_gh or html_is_gh

    title = " ".join(parser.h1_text).strip() or " ".join(parser.title_text).strip() or parser.meta.get("og:title")
    employer = None
    job_title = title
    if title and " at " in title:
        # "Role at Company" pattern
        parts = title.rsplit(" at ", 1)
        if len(parts) == 2:
            job_title, employer = parts[0].strip(), parts[1].strip()

    board = None
    if application_url:
        m = re.search(r"greenhouse\.io/([^/?#]+)", application_url)
        if m:
            board = m.group(1)

    fields: list[DetectedField] = []
    for f in parser.fields:
        # Re-resolve label from for= map
        label = f.get("label") or ""
        if not label and f.get("name"):
            label = f["name"].replace("_", " ")
        df = _classify_detected(label, f.get("type") or "text", f.get("name") or "")
        df.is_required = bool(f.get("required"))
        fields.append(df)

    return _finalize(GreenhouseDetectionResult(
        is_greenhouse=is_gh,
        detection_mode="html",
        employer=employer,
        job_title=job_title,
        application_url=application_url,
        board_token=board,
        fields=fields,
    ))


def detect_from_greenhouse_api_json(data: dict | str, *, application_url: Optional[str] = None) -> GreenhouseDetectionResult:
    if isinstance(data, str):
        data = json.loads(data)
    board = data.get("board")
    absolute = data.get("absolute_url") or application_url
    fields: list[DetectedField] = []

    def add_question(q: dict, prefix: str = ""):
        label = q.get("label") or ""
        qid = str(q.get("name") or q.get("id") or label)
        required = bool(q.get("required"))
        fields_list = q.get("fields") or [{"type": q.get("type") or "input_text", "name": qid}]
        for fld in fields_list:
            ftype = fld.get("type") or "input_text"
            name = str(fld.get("name") or qid)
            df = _classify_detected(label, ftype, name)
            df.is_required = required
            if ftype in ("multi_value_single_select", "multi_value_multi_select"):
                opts = fld.get("values") or q.get("values") or []
                df.options = [str(o.get("label") if isinstance(o, dict) else o) for o in opts][:50]
            fields.append(df)

    for q in data.get("questions") or []:
        add_question(q)
    for q in data.get("location_questions") or []:
        add_question(q, "location")
    for q in data.get("compliance") or []:
        add_question(q, "compliance")

    demo = data.get("demographic_questions")
    if demo:
        for q in (demo.get("questions") if isinstance(demo, dict) else []) or []:
            df = _classify_detected(q.get("label") or "demographic", "demographic", str(q.get("id") or "demo"))
            df.classification = "sensitive_question"
            df.normalized_field_name = None
            fields.append(df)

    return _finalize(GreenhouseDetectionResult(
        is_greenhouse=True,
        detection_mode="greenhouse_api_json",
        employer=board,
        job_title=data.get("title"),
        application_url=absolute,
        board_token=board,
        fields=fields,
    ))


def _finalize(result: GreenhouseDetectionResult) -> GreenhouseDetectionResult:
    result.filled_any_fields = False
    result.submitted = False
    for f in result.fields:
        if f.classification == "supported" and f.normalized_field_name:
            if f.normalized_field_name not in result.supported_fields:
                result.supported_fields.append(f.normalized_field_name)
        elif f.classification == "sensitive_question":
            result.sensitive_fields.append(f.field_label or f.external_field_key)
        elif f.classification == "custom_question":
            result.custom_fields.append(f.field_label or f.external_field_key)
        elif f.classification == "legal_attestation":
            result.legal_fields.append(f.field_label or f.external_field_key)
        elif f.classification in ("unsupported", "unknown"):
            result.unsupported_fields.append(f.field_label or f.external_field_key)
        if f.is_upload:
            result.upload_controls.append(f.field_label or f.external_field_key)
        if f.is_required:
            result.required_fields.append(f.field_label or f.external_field_key)
    if not result.is_greenhouse:
        result.warnings.append("Page/URL does not look like Greenhouse")
    return result


def detect_fixture_file(path: str | Path) -> GreenhouseDetectionResult:
    p = Path(path)
    text = p.read_text(encoding="utf-8")
    if p.suffix.lower() == ".json":
        return detect_from_greenhouse_api_json(json.loads(text))
    return detect_from_html(text)
